### PR TITLE
[SPARK-34008][BUILD] Upgrade derby to 10.14.2.0

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR upgrades `derby` to `10.14.2.0`.

You can check the major changes from the following URLs.

* 10.13.1.1 http://svn.apache.org/repos/asf/db/derby/code/tags/10.13.1.1/RELEASE-NOTES.html
* 10.14.1.0 http://svn.apache.org/repos/asf/db/derby/code/tags/10.14.1.0/RELEASE-NOTES.html
* 10.14.2.0 http://svn.apache.org/repos/asf/db/derby/code/tags/10.14.2.0/RELEASE-NOTES.html

### Why are the changes needed?

It seems to be the final release which supports `JDK8` as the minimum required version.
After `10.15.1.3`, the minimum required version is `JDK9`.
https://db.apache.org/derby/

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @sunchao FYI

**Comment by maropu** [comment]
Could you put major change logs from `10.12.1.1` to `10.14.2.0` in the PR description?

**Review by srowen — COMMENTED** [review]
Seems OK; is there any reason to believe this changes behavior and/or picks up fixes and improvements that Spark could use?

File: pom.xml [hunk 1]
```diff
@@ -134,7 +134,7 @@
     <hive.version.short>2.3</hive.version.short>
     <!-- note that this should be compatible with Kafka brokers version 0.10 and up -->
     <kafka.version>2.6.0</kafka.version>
-    <derby.version>10.12.1.1</derby.version>
```
Comment (dongjoon-hyun) [inline pom.xml:137]: Could you add a comment, `After 10.15.1.3, the minimum required version is JDK9`, here?

**Comment by sarutak** [comment]
@srowen Those changes contains some bug fixes which affects `10.12.2.0`. Though I don't know whether those bugs affect Spark but it's for the stability.
The changes also include some incompatibilities but I don't think these incompatibilities affect Spark ( please double check just in case).
http://svn.apache.org/repos/asf/db/derby/code/tags/10.13.1.1/RELEASE-NOTES.html#Note%20for%20DERBY-6852
http://svn.apache.org/repos/asf/db/derby/code/tags/10.14.2.0/RELEASE-NOTES.html#Note%20for%20DERBY-6987
http://svn.apache.org/repos/asf/db/derby/code/tags/10.14.2.0/RELEASE-NOTES.html#Note%20for%20DERBY-6986

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM for Apache Spark 3.2.0. Thank you, @sarutak and @srowen and @maropu and @HyukjinKwon .
Merged to master.

**Pull request closed without merging**